src/models/read-receipt.ts

```diff
@@ -95,14 +96,74 @@ export abstract class ReadReceipt<
 
     /**
      * Get the ID of the event that a given user has read up to, or null if we
-     * have received no read receipts from them.
+     * have received no read receipts from them, or if the receipt we have
+     * points at an event we don't have, or if the thread ID in the receipt does
+     * not match the thread root of the referenced event.
+     *
      * @param userId - The user ID to get read receipt event ID for
      * @param ignoreSynthesized - If true, return only receipts that have been
      *                                    sent by the server, not implicit ones generated
      *                                    by the JS SDK.
-     * @returns ID of the latest event that the given user has read, or null.
+     * @returns ID of the latest existing event that the given user has read, or null.
      */
     public getEventReadUpTo(userId: string, ignoreSynthesized = false): string | null {
+        // Find what the latest receipt says is the latest event we have read
+        const latestReceipt = this.getLatestReceipt(userId, ignoreSynthesized);
+
+        if (!latestReceipt) {
+            return null;
+        }
+
+        return this.receiptPointsAtConsistentEvent(latestReceipt) ? latestReceipt.eventId : null;
+    }
+
+    /**
+     * Returns true if the event pointed at by this receipt exists, and its
+     * threadRootId is consistent with the thread information in the receipt.
+     */
+    private receiptPointsAtConsistentEvent(receipt: WrappedReceipt): boolean {
+        // If the receipt points at a non-existent event, it's not consistent
+        const event = this.findEventById(receipt.eventId);
+        if (!event) {
+            logger.warn(`Ignoring receipt for missing event with id ${receipt.eventId}`);
+            return false;
+        }
```

Comment:
Doesn't this automatically do the filtering-by-thread-id stuff? `Thread.findEventById` will return null if the event is in another thread?